gensim/models/base_any2vec.py

```diff
@@ -459,25 +469,34 @@ def __str__(self):
             self.__class__.__name__, len(self.wv.index2word), self.vector_size, self.alpha
         )
 
-    def build_vocab(self, sentences, update=False, progress_per=10000, keep_raw_vocab=False, trim_rule=None, **kwargs):
+    def build_vocab(self, sentences, multistream=False, workers=None, update=False, progress_per=10000,
+                    keep_raw_vocab=False, trim_rule=None, **kwargs):
         """Build vocabulary from a sequence of sentences (can be a once-only generator stream).
         Each sentence is a iterable of iterables (can simply be a list of unicode strings too).
 
         Parameters
         ----------
-        sentences : iterable of iterables
+        sentences : {iterable of iterables, list or tuple of iterable of iterables}
```

Comment:
I'd prefer not to override the same parameter with a different type (with the type dependant on an external flag).

It's already complicated enough; `iterable of iterables, or a list or tuple of iterable of iterables` is no longer humanly parseable.

My original suggestion was to have it a sequence of iterables *always*. If backward compatibility is hard to achieve (is it?), keep the legacy parameter `sentences` (maybe deprecate it in time) and promote a new parameter `multistream`. If backward compatibility can be achieved transparently, then just keep using `sentences`.

Reply:
Correct me if I got you wrong, but you propose smth like this:

`Word2Vec(sentences=None, input_streams=None, ...)` (instead of old `Word2Vec(sentences=None, ...)` )

with the following logic:
1. both `sentences` and `input_streams` can't be passed
2. if `sentences` is passed, then do `input_streams = [sentences]` and go to (3)
3. if `input_streams` is passed, then perform multistream mode


Am I correct about your logic?

Reply:
Yes. I originally thought users would only pass `sentences`, and we decide automatically whether it's a single stream (legacy, promote to `input_stream = [sentences]`) or already multiple streams (new-style). And keep only that single parameter.

But introducing extra `input_streams` parameter may be a cleaner option. More explicit.

Reply:
Okay, I'll implement the logic in this way, sounds good

Reply:
@piskvorky Currently, in a majority of methods (almost all except constructor) the `sentences` parameter is required, not optional. If we introduce new `input_streams` param, both of them need to be optional.

This requires us to do some extra checks e.g. `assert sentences is not None or input_streams is not None` in all of the methods. It sounds a bit dirty

What do you think about that?

Reply:
Do you see any ambiguity/problems with detecting "legacy" (single stream) `sentences` automatically?

Reply:
• `sentences` is like a `[['first', 'sentence'], ['second', 'sentence']]`
• `input_streams` is like a `[ [ [ 'first', 'sentence1' ], [ 'second', 'sentence1' ] ] , [ ['first', 'sentence2' ], [ 'second', 'sentence2' ] ] ]`

So, to determine the nesting level (whether its like `sentences` or like `input_streams`) it's necessary to obtain at least the first element from this argument (execute `next(...)`, maybe twice), if this is okay then no problems. 

Pseudocode for this logic:

```
def is_multistream(sentences):
  nesting_level = 0
  current_elem = sentences
  while not isinstance(current_elem, string_types):
    current_elem = next(current_elem)
    nesting_level += 1
  if nesting_level == 2:
      return False  # this is single stream like [['first', 'sentence'], ...]
  elif nesting_level == 3:
      return True   # this is multistream like [ [['first', 'sentence'], ...], ...]
  else:
      raise RuntimeError('this is not possible')
```

Reply:
Well, we'd have to put the "peeked" items back of course. We cannot just discard them :)

We're already doing something similar in `utils.is_corpus()`.

Question is, is it possible to do something similar here as well? What is the risk of unintended consequences, such as accidentally materializing some huge corpus?

Because it is a bit hackish, so the tradeoff (pros and cons) between having an extra explicit parameter vs being simpler for user is not clearcut to me.

Reply:
> Question is, is it possible to do something like that here as well? What is the risk of unintended consequences, such as accidentally materializing some huge streamed object in RAM?

I don't see any problems here. This object will be "materialized" in the code, anyway, when we will be iterate over the data.

> Because the peeking is a bit hackish, so the tradeoff (pros and cons) between having an extra explicit parameter vs being simpler for user is not clearcut to me.

Agree, maybe two parameters `sentences=None, input_streams=None` + check that one of them is not None, is more appropriate?

Reply:
Yeah, alright.

Btw how will N streams interact with M workers? Or do we force N==M?

That could make life harder for the most common use-case, where user has only one stream, N=1. Then they couldn't train in parallel at all.

Reply:
N streams are joined into one https://github.com/RaRe-Technologies/gensim/pull/2078/files#diff-ce0adce9f4ad6c3715b44dced29c2759R219

Reply:
Making sure I understand:
1. if M==1, we use the existing algorithm global queue and merge all streams into one, forcing N==1
2. if M > 1, we use an algorithm that does no queuing. And the N streams are distributed among M workers how?

Reply:
In this pull request, there are no training optimizations at all, only during dictionary building stage. We still use the existing "global queue" algorithm for any number of input_streams (merging them into one).

I don't know yet the way we will go, because currently I don't have positive experiments with multistream training.

Reply:
Dictionary building is a part of training. I'm still unclear what happens when `N != M` (or do we force `N==M` always?).

Reply:
No, we don't force anything. I just use `multiprocessing.Pool`, put `len(input_streams)` tasks in it and that's all.

Reply:
https://github.com/RaRe-Technologies/gensim/pull/2078/files#diff-673fdc31e9aae23291039b143f451b9eR1236 here is the exact logic I use when creating the pool

Reply:
Small clarification: this is **only** for dictionary building (see PR title), training process stay as is, the goal of PR is parallelized vocab building only.

Reply:
OK, so the task distribution is managed by Pool (at least during this training stage), got it.